### PR TITLE
BUG: Don't generate BPF for psuedo-syscalls

### PR DESCRIPTION
**Pull request by drakenclimber**
This patchset:
- Removes pseudo-syscalls from the output BPF filter
- Fixes a couple potential bugs in the x32 and arm code where we were doing subtraction/addition rather than bitwise operations
- Updates two tests to work with the BPF/pseudo-syscalls change

### CODE REVIEW

**Comment by drakenclimber** [comment]
Fixes Issue #252

File: src/gen_bpf.c [hunk 1]
```diff
@@ -1291,6 +1291,50 @@ static int _gen_bpf_insert(struct bpf_state *state, struct bpf_instr *instr,
 	return rc;
 }
 
+/**
+ * Omit syscall from the BPF filter?
```
Comment (pcmoore) [inline src/gen_bpf.c:1295]: Big time nitpick, but typically the one-liner summaries are statements not questions, e.g. "Decide if we need to omit the syscall from the BPF filter".

File: src/gen_bpf.c [hunk 2]
```diff
@@ -1532,11 +1576,17 @@ static int _gen_bpf_syscalls(struct bpf_state *state,
...
-					   bintree_levels, db->syscall_cnt,
+					   bintree_levels, syscall_cnt,
 					   &empty_cnt);
 		if (rc < 0)
 			goto out;
+
+		/* reset the syscall_cnt variable because later in this
+		 * function it's used as a counter
+		 */
+		syscall_cnt = 0;
```
Comment (pcmoore) [inline src/gen_bpf.c:1589]: We can drop this special handling if we move the `syscall_cnt` initialization to 0 from the declaration line to right about the for-loop where it is first used.

File: src/gen_bpf.c [hunk 3]
```diff
@@ -1555,6 +1605,8 @@ static int _gen_bpf_syscalls(struct bpf_state *state,
 	for (s_iter = s_tail; s_iter != NULL; s_iter = s_iter->pri_prv) {
 		if (!s_iter->valid)
 			continue;
+		if (_skip_syscall(state, s_iter))
```
Comment (pcmoore) [inline src/gen_bpf.c:1608]: Same as above, might as well combine this with the valid check.

File: src/gen_bpf.c (hunk 1 above)
Comment (drakenclimber) [reply src/gen_bpf.c:1295]: :) Can you tell I was in a Java frame of mind?  Thanks, I like your wording better.

File: src/gen_bpf.c [hunk 4]
```diff
@@ -1291,6 +1291,50 @@ static int _gen_bpf_insert(struct bpf_state *state, struct bpf_instr *instr,
...
+ * @param syscall syscall being tested
+ * @return true if syscall is to be skipped, false otherwise
+ */
+static inline bool _skip_syscall(struct bpf_state *state,
+				 struct db_sys_list *syscall)
+{
+	/* psuedo-syscalls should not be added to the filter unless explicity
+	 * requested via SCMP_FLTATR_API_TSKIP
+	 */
+	if ((syscall->num & 0x80000000) &&
```
[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/gen_bpf.c:1306]: I'm glad you brought this up.  I copy-pasta'ed this from the proposed patchset you wrote [here](https://github.com/seccomp/libseccomp/issues/252#issuecomment-644154303), and I was curious why.  I'll definitely switch to using `<` as that reads so much better.

[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/gen_bpf.c:1306]: Looks like you used the bitwise operator because num is defined as an unsigned int.  I'll use `<` and a cast to an int.

**Comment by pcmoore** [comment]
Merged at HEAD 2f3ae1fe7eaa1195e29979ae341bfc288bc69ee5, thanks!

**Comment by vt-alt** [comment]
Thanks!

**Pull request closed without merging**